### PR TITLE
feat: Add tfidf

### PR DESCRIPTION
**Pull request by digriffiths**
added tfidf vectorizer, changed hybrid layer to accomodate, user can define which sparse encoder they use

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -19,6 +19,7 @@ cohere = "^4.32"
 numpy = "^1.25.2"
 pinecone-text = "^0.7.0"
 colorlog = "^6.8.0"
+scikit-learn = "^1.3.2"
```
Comment (jamescalam) [inline pyproject.toml:22]: `scikit-learn` is a very heavy library — I'd rather we avoid this at all costs, should be possible to implement TFIDF with pure numpy

**Review by jamescalam — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 1 above)
Comment (digriffiths) [reply pyproject.toml:22]: Thanks James, I'll get that changed.

Comment (simjak) [reply pyproject.toml:22]: image

**Review by italianconcerto — APPROVED** [review]

**Review by jamescalam — APPROVED** [review]
awesome work @digriffiths well done! Merging

**Pull request merged**